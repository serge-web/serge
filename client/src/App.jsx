import { Component } from 'react'
import { Provider } from 'react-redux'
import { MuiThemeProvider as ThemeProvider } from '@material-ui/core/styles'
import theme from './themes/serge'
// import Notifications from './Components/Notifications'
// import Version from './Views/Version'
import RouterMain from './Components/Router/RouterMain'
import Store from './Store/Store'
// import '@serge/themes/App.scss'
import './themes/App.scss'


class App extends Component {
  componentDidMount () {
    document.title = 'Serge'
  }
  
  render () {
    return (
      <Provider store={Store}>
        <ThemeProvider theme={theme}>
          <RouterMain />
          {/* <Notifications /> */}
          {/* <Version /> */}
        </ThemeProvider>
      </Provider>
    )
  }
}

export default App
