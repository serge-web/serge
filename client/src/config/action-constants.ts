export const ASYNC_SUCCESS = 'ASYNC_SUCCESS'
export const ASYNC_ERROR = 'ASYNC_ERROR'
export const ASYNC_LOADING = 'ASYNC_LOADING'
export const NORMAL_ACTION = 'NORMAL_ACTION'
export const DB_MESSAGE_STATUS = 'DB_MESSAGE_STATUS'
export const DB_MESSAGE_CREATION_LOADING = 'DB_MESSAGE_CREATION_LOADING'
export const DB_MESSAGE_GET_LOADING = 'DB_MESSAGE_GET_LOADING'
export const DB_SAVE_MESSAGES = 'DB_SAVE_MESSAGES'
export const DB_MESSAGE_TYPES_GET = 'DB_MESSAGE_TYPES_GET'
export const DB_MESSAGE_TYPES_SAVED = 'DB_MESSAGE_TYPES_SAVED'
export const SET_CURRENT_VIEW_FROM_URI = 'SET_CURRENT_VIEW_FROM_URI'
export const DB_MESSAGE_SAVED = 'DB_MESSAGE_SAVED'
export const DB_MESSAGES_GET = 'DB_MESSAGES_GET'
export const DB_RETURNED_MESSAGE = 'DB_RETURNED_MESSAGE'
export const RESET_MESSAGE_PREVIEW = 'RESET_MESSAGE_PREVIEW'
export const SET_SELECTED_SCHEMA = 'SET_SELECTED_SCHEMA'
export const SET_PREVIEW_SCHEMA = 'SET_PREVIEW_SCHEMA'
export const OPEN_MODAL = 'OPEN_MODAL'
export const CLOSE_MODAL = 'CLOSE_MODAL'
export const SET_CURRENT_GAME_SETUP_TAB = 'SET_CURRENT_GAME_SETUP_TAB'
export const SET_GAME_SETUP_DATA = 'SET_GAME_SETUP_DATA'
export const ALL_WARGAME_NAMES_SAVED = 'ALL_WARGAME_NAMES_SAVED'
export const SET_CURRENT_WARGAME = 'SET_CURRENT_WARGAME'
export const ADD_NEW_FORCE = 'ADD_NEW_FORCE'
export const SET_SELECTED_FORCE = 'SET_SELECTED_FORCE'
export const ADD_NEW_CHANNEL = 'ADD_NEW_CHANNEL'
export const SET_SELECTED_CHANNEL = 'SET_SELECTED_CHANNEL'
export const DELETE_SELECTED_CHANNEL = 'DELETE_SELECTED_CHANNEL'
export const GAME_TITLE_INVALID = 'GAME_TITLE_INVALID'
export const ADD_NEW_RECIPIENT = 'ADD_NEW_RECIPIENT'
export const ADD_NOTIFICATION = 'ADD_NOTIFICATION'
export const ADD_PLAYER_LOG = 'ADD_PLAYER_LOG'
export const HIDE_NOTIFICATION = 'HIDE_NOTIFICATION'
export const UPDATE_RECIPIENT = 'UPDATE_RECIPIENT'
export const REMOVE_RECIPIENT = 'REMOVE_RECIPIENT'
export const SET_CURRENT_WARGAME_PLAYER = 'SET_CURRENT_WARGAME_PLAYER'
export const SET_FORCE = 'SET_FORCE'
export const SET_ROLE = 'SET_ROLE'
export const SET_FILTERED_CHANNELS = 'SET_FILTERED_CHANNELS'
export const SET_CHANNEL = 'SET_CHANNEL'
export const SET_MESSAGE_SCHEMA = 'SET_MESSAGE_SCHEMA'
export const SET_ALL_MESSAGES = 'SET_ALL_MESSAGES'
export const POPULATE_MESSAGE_TYPES_DB = 'POPULATE_MESSAGE_TYPES_DB'
export const POPULATE_WARGAMES_DB = 'POPULATE_WARGAMES_DB'
export const SET_UNSAVED_STATE = 'SET_UNSAVED_STATE'
export const SET_WARGAME_TITLE = 'SET_WARGAME_TITLE'
export const UPDATE_CHANNEL_NAME = 'UPDATE_CHANNEL_NAME'
export const SET_TAB_SAVED = 'SET_TAB_SAVED'
export const SET_TAB_UNSAVED = 'SET_TAB_UNSAVED'
export const ADD_ROLE_TO_FORCE = 'ADD_ROLE_TO_FORCE'
export const REMOVE_ROLE = 'REMOVE_ROLE'
export const NEXT_GAME_TURN = 'NEXT_GAME_TURN'
export const UPDATE_ROLE_NAME = 'UPDATE_ROLE_NAME'
export const ADD_ICON = 'ADD_ICON'
export const SET_EXPORT_WARGAME = 'SET_EXPORT_WARGAME'
export const CREATE_EXPORT_ITEM = 'CREATE_EXPORT_ITEM'
export const SHOW_HIDE_OBJECTIVES = 'SHOW_HIDE_OBJECTIVES'
export const UPDATE_MESSAGE_STATE = 'UPDATE_MESSAGE_STATE'
export const SET_FEEDBACK_MESSAGES = 'SET_FEEDBACK_MESSAGES'
export const SET_ALL_TEMPLATES_PLAYERUI = 'SET_ALL_TEMPLATES_PLAYERUI'
export const SET_LATEST_WARGAME_MESSAGE = 'SET_LATEST_WARGAME_MESSAGE'
export const SET_LATEST_FEEDBACK_MESSAGE = 'SET_LATEST_FEEDBACK_MESSAGE'
export const OPEN_MESSAGE = 'OPEN_MESSAGE'
export const MARK_UNREAD = 'MARK_UNREAD'
export const CLOSE_MESSAGE = 'CLOSE_MESSAGE'
export const MARK_ALL_AS_READ = 'MARK_ALL_AS_READ'
export const MARK_ALL_AS_UNREAD = 'MARK_ALL_AS_UNREAD'
export const SET_FORCE_COLOR = 'SET_FORCE_COLOR'
export const LOAD_SERGE_GAME_INFO = 'LOAD_SERGE_GAME_INFO'
export const LOGIN_ADMIN = 'LOGIN_ADMIN'
export const OPEN_TOUR = 'OPEN_TOUR'
export const SET_SERVER_STATUS = 'SET_SERVER_STATUS'
export const LOADER = 'LOADER'
