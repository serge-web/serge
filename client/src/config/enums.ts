/**
 * list the possible phases in a wargame
 */
export enum Phase {
  /* players are planning their next turn
   */
  Planning = 'planning',
  /** umpire is resolving planned turns */
  Adjudication = 'adjudication'
}

/** increasing permissions in a collaborative editing channel
 */
export enum CollaborativePermission {
  /** can */
  CannotCollaborate,
  CanEdit,
  CanSubmitForReview,
  CanApprove,
  CanRelease,
  CanUnClaim
}

/** the initial state for new messages in a collab edit channel */
export enum InitialStates {
  PENDING_REVIEW = 'PENDING_REVIEW',
  UNALLOCATED = 'UNALLOCATED'
}

/** types of special channel */
export enum SpecialChannelTypes {
  CHANNEL_COLLAB = 'collab',
  CHANNEL_CHAT = 'chat',
  CHANNEL_CORE_MAPPING = 'core mapping'
}

/** types of extra column to be shown in collab editing */
export enum SpecialChannelColumns {
  LOCATION = 'location'
}

/** the phases a v3 collaborative message goes through before
 * being approved to send
 */
export enum CollaborativeMessageStates {
  /**
   * message is waiting to be allocated to owner
   */
  Unallocated = 'Unallocated',
  /**
   * non collab-editing (e.g. RFI) document being edited
   * //TODO: remove this when we retire RFI status board.
   */
  InProgress = 'In progress',
  /**
   * owner finished editing, waiting for review
   */
  PendingReview = 'Pending review',
  /**
   * approved and released
   */
  Released = 'Released',
  /** 
   * document cancelled 
   */
  Closed = 'Closed',
}

/** the commands that can be appled to a collaborative message
 */
export enum CollaborativeMessageCommands {
  SendForReview = 'Send for review',
  TakeOwnership = 'Take ownership',
  ReOpen = 'Re-open',
  Close = 'Close',
  Release = 'Release',
  RequestChanges = 'Request changes'
}

/** the ways in which turn numbers may be displayed */
export enum TurnFormats {
  // 0, 1, 2, 3
  Natural = 'Natural',
  // 0.1, 0.2, 1.1, 1.2
  TurnPairNumbers = 'Turn-Pair-Numbers',
  // 0.a, 0.b, 1.a, 1.b
  TurnPairLetters = 'Turn-Pair-Letters'
}

export enum AdminTabs {
  Overview = 'overview',
  Forces = 'forces',
  Channels = 'channels',
  Templates = 'templates'
}

export enum TemplateTab {
  Preview = 'Preview Form',
  Visual = 'Visual Form Builder'
}
