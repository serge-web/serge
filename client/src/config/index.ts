/***
* This file is autogenerated, please do not edit it.
* To generate an updated version, please run `yarn genreg`
***/

export * from './globals'
export * from './routes'
export * from './enums'
export * from './action-constants'
