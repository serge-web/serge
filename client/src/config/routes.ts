export const ADMIN_ROUTE = '/serge/admin'
export const MESSAGE_TEMPLATE_ROUTE = '/serge/messageTemplates'
export const MESSAGE_CREATOR_BASE_ROUTE = '/serge/messageCreator'
export const CREATE_TEMPLATE_ROUTE = '/create/template'
export const EDIT_TEMPLATE_ROUTE = '/edit/template'
export const CREATE_MESSAGE_ROUTE = '/create/message'
export const EDIT_MESSAGE_ROUTE = '/edit/message'
export const GAME_SETUP_ROUTE = '/serge/gameSetup'
export const WELCOME_SCREEN_EDIT_ROUTE = '/serge/editWelcomeScreen'
export const REST_API_ROUTE = 'api-docs/'
export const EXPORT_ROUTE = '/serge/export'
export const EXPORT_MESSAGES_SUBROUTE = '/messages'
export const EXPORT_FORCES_SUBROUTE = '/foeces'
export const EXPORT_PRINT_SUBROUTE = '/print/:id'
export const PLAYERUI_ROUTE = '/serge/player'
export const DEMO_ROUTE = '/serge/demo'
