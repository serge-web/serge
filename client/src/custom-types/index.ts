/***
* This file is autogenerated, please do not edit it.
* To generate an updated version, please run `yarn genreg`
***/

export * from './channel-data.d'
export type { default as ChannelUI } from './channel-ui.d'
export type { default as Channel } from './channel.d'
export * from './export-items-actions.d'
export * from './export-items.d'
export type { default as ForceData } from './force-data.d'
export type { default as ForceOption } from './force-option.d'
export type { default as ForceRole } from './force-role.d'
export type { default as Force } from './force.d'
export * from './global.d'
// export { default as MapPostBack } from './map-postback-data.d'
export type { default as MarkerPerception } from './marker-perception.d'
export type { default as MessageTypes, TemplateBodysByKey } from './message-types.d'
export * from './message.d'
export type { SetCurrentViewFromURI } from './messages-action.d'
export * from './participant.d'
export * from './player-log-actions.d'
export * from './player-log.d'
export * from './player-ui-actions.d'
export type { default as PlayerUi, PlayerUiChannels, PlayerUiChatChannel } from './player-ui.d'
export type { default as WargamesState } from './WargamesState.d'
export type { default as Role } from './role.d'
export * from './root-state'
export type { default as SetWargameMessage } from './set-wargame-message.d'
export type { default as Template, TemplateBody, TemplateData } from './template.d'
export * from './turn-length.d'
export * from './wargame-action.d'
export type { default as WargameData } from './wargame-data.d'
export type { default as WargameList } from './wargame-list.d'
export type { default as WargameOverview } from './wargame-overview.d'
export type { default as Wargame } from './wargame.d'
export * from './core-mapping.d'
export * from './turn-period.d'
