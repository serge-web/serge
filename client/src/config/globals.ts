import ExpiredStorage from 'expired-storage'
import { Wargame } from 'src/custom-types'
// import { wargameSettings } from 'src/config'
import uniqId from 'uniqid'
import moment from 'moment'

export const DEFAULT_SERVER = 'Nelson'
export const DEFAULT_PORT = '8080'

export const MSG_STORE = 'messages'
export const MSG_TYPE_STORE = 'message_types'
export const SERGE_INFO = 'serge_info'
export const CHAT_CHANNEL_ID = 'game-admin'

export const PLANNING_PHASE = 'planning'
export const ADJUDICATION_PHASE = 'adjudication'

export const expiredStorage = new ExpiredStorage()
export const LOCAL_STORAGE_TIMEOUT = 2592000 // one month

export const UMPIRE_FORCE = 'umpire'

export const LOCATION_PENDING = 'LocationPending' // special state where platforms can be moved at turn zero

// series of constants used for `messageType` when sending map events

// series of constants used for `messageType` when sending custom messages
export const CUSTOM_MESSAGE = 'CustomMessage'

// series of constants used for `messageType` when sending chat messages
export const CHAT_MESSAGE = 'ChatMessage'

// series of constants used for `messageType` when sending feedback
export const FEEDBACK_MESSAGE = 'FeedbackMessage'

// series of constants used for `messageType` when using Counter message for COA and RFI messages
export const COUNTER_MESSAGE = 'CounterMessage'

// mapping messages
export const MAPPING_MESSAGE = 'MappingMessage'
export const MAPPING_MESSAGE_DELTA = 'MappingMessageDelta'

// series of constants used for `messageType` when sending system messages
// an INFO_MESSAGE is an update to the wargame document
export const INFO_MESSAGE = 'InfoMessage'
export const INFO_MESSAGE_CLIPPED = 'InfoMessageClipped'

// return states of adjudication form
export const PLAN_ACCEPTED = 'accepted'
export const PLAN_REJECTED = 'rejected'

// types of UI interaction
export const PLAIN_INTERACTION = 'plain-interaction'
export const CHANGE_TAB_INTERACTION = 'change-tab-interaction'
export const MESSAGE_SENT_INTERACTION = 'msg-sent-interaction'
export const MESSAGE_READ_INTERACTION = 'msg-read-interaction'
export const MESSAGE_UNREAD_INTERACTION = 'msg-unread-interaction'
export const MAP_ANNOTATION_READ_INTERACTION = 'map-anno-read-interaction'

// export item types
export const EXPORT_ITEM_MESSAGES = 'messages'
export const EXPORT_ITEM_FORCES = 'forces'

// names of special channels
export const CHANNEL_RFI_STATUS = 'rfis'

// types of channel
export const CHANNEL_CHAT = 'ChannelChat'
export const CHANNEL_CUSTOM = 'ChannelCustom'
export const CHANNEL_COLLAB = 'ChannelCollab'
export const CHANNEL_MAPPING = 'ChannelMapping'

// types of participant
export const PARTICIPANT_CHAT = 'ParticipantChat'
export const PARTICIPANT_CUSTOM = 'ParticipantCustom'
export const PARTICIPANT_COLLAB = 'ParticipantCollab'
export const PARTICIPANT_MAPPING = 'ParticipantMapping'

// Chat template ID
export const CHAT_MESSAGE_TEMPLATE_ID = 'k16eedkl'

// value to use in DataTable filters for value not present
export const EMPTY_CELL = '[Empty]'

// name of property used for storing time/type of last activity

// marker character to tell updateMarker handler that the update is coming 
// from the marker form - and the maker may have been separately dragged around
export const FLAG_MARKER = '!'

// NOTE: time period to wait if server returns an error. One frequent cause of error
// during development is that the server is stopped.  We're introducing a
// throttle value to prevent the browser going into a race condition
// as it sends 1000s of requests to the server
export const ERROR_THROTTLE = 3000

// Nov 2019. Ian modified the server path to use the
// current URL, so we can use Heroku to provide
// review instances of the app.  In these
// review instances, we can't predict the URL, so
// were failing CORS test

const { hostname, protocol, href, origin } = window.location

export const baseUrl = () => {
  const host = (new URL(href)).searchParams.get('host')

  // NOTE: for all non-heroku deployments, we need to append the port number
  const portDetails = hostname.toLowerCase().indexOf('herokuapp') !== -1 ? '' : ':' + DEFAULT_PORT

  const res = host || `${protocol}//${hostname}` + portDetails

  return res
}

export const serverPath = (
  // Removed G_CONFIG tempoararily.
  // window.G_CONFIG.REACT_APP_SERVER_PATH || process.env.REACT_APP_SERVER_PATH || baseUrl() + '/'
  process.env.REACT_APP_SERVER_PATH || baseUrl() + '/'
).replace(/\/?$/, '/')

export const databasePath = `${serverPath}db/`
export const iconUploaderPath = `${serverPath}saveIcon`
export const hiddenPrefix = '_#_'

export const clearAll = 'clearAll'
export const allDbs = 'allDbs'
export const playerlogs = 'playerlogs'
// Note: On heroku we don't use the additional port for the socket, we use the plain origin
export const socketPath = origin.toLowerCase().indexOf('herokuapp') !== -1 ? origin : origin.replace(/3000|8080/g, '4000')
export const replicate = 'replicate/'
export const deletePath = 'delete/'
/** name of the document used to store the initial wargame definition */
export const wargameSettings = 'initial_wargame'
// default name for role name
export const NEW_ROLE = 'New Role'

/** flag to indicate a role can control all assets */

// 
/** there has been some user interaction, so log the current time
 * The storage we're using is shared across browser tabs, so previously
 * one session would overwrite the history for another. So, we prepend the
 * activity type with the role name - to let one player use multiple tabs.
 * @param role - current role id
 * @param event - name of event that just happened
 */
export const UNSENT_CHAT_MESSAGE_TYPE = 'chat'
export const UNSENT_SELECT_BY_DEFAULT_VALUE = 'BY_DEFAULT_VALUE'

export const STORYBOOK_ROUTE = './storybook'

// Interval for check server heartbeats
export const SERVER_PING_INTERVAL = 20000

export const headers = { // +
  'Content-Type': 'application/json',
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'POST',
  'Access-Control-Allow-Headers': 'Authorization, Lang'
}

export const defaultGameInfo = { // +
  imageUrlSet: false,
  imageUrl: '/default_img/sergeDefault.png',
  title: 'Serge',
  description: `You have arrived at the Development Centre Gaming Facility.\n
  You will use this web-based application to interact with players from other forces, together with the umpires in the White Cell.\n
  At any point during your time here you can submit insights via the Insights button at the top-right of the gaming page.\n
  These insights could relate to the current doctrine being explored, the performance of your force, or how the game is being organised / facilitated.\n
  Thanks in advance for your participation.\n
  Maj Duncan Dare, PO1 Gaming`,
  showAccessCodes: true
}

export const forceTemplate = { // +
  name: '',
  uniqid: '',
  overview: 'An overview written here..',
  roles: [
    {
      name: 'CO',
      roleId: `p${uniqId.time()}`,
      isGameControl: false,
      isObserver: false,
      isInsightViewer: false
    }
  ],
  iconURL: serverPath + 'default_img/forceDefault.png',
  color: '#3dd0ff',
  umpire: false,
  dirty: false
}

export const templateBody = {
  lastUpdated: '2019-09-30T12:37:26.705Z',
  title: 'Chat',
  details: {
    type: 'object',
    properties: {
      content: {
        type: 'string',
        format: 'textarea',
        options: {
          inputAttributes: {
            placeholder: 'type the text'
          }
        }
      }
    },
    title: 'Chat',
    
    format: 'grid'
  },
  completed: false,
  _id: 'k16eedkl',
  _rev: '1-09ab7a18ff677cec5d9a56f02a45788d'
}

export const umpireForceTemplate = { // +-
  name: 'White',
  uniqid: 'umpire',
  overview: 'Umpire force.',
  roles: [
    {
      name: 'Game Control',
      roleId: `p${uniqId.time()}`,
      isGameControl: true,
      isObserver: true,
      isInsightViewer: true
    }
  ],
  iconURL: serverPath + 'default_img/umpireDefault.png',
  color: '#FCFBEE',
  umpire: true,
  dirty: false
}

export const channelTemplate = { // +
  name: '',
  uniqid: '',
  participants: []
}

export const dbDefaultPlaylogSettings = { // +-
  wargame: 'missing',
  role: 'missing',
  activityType: {
    aType: 'pending'
  },
  activityTime: '',
  _id: new Date().toISOString(), 
  isOpen: false
}

export const dbDefaultSettings: Wargame = { // +
  _id: wargameSettings,
  wargameTitle: '',
  name: '',
  data: {
    overview: {
      name: 'Overview - settings',
      gameDescription: '',
      // spatialRepresentation: '',
      gameTurnTime: {
        unit: 'millis',
        millis: 43200000
      },
      realtimeTurnTime: 300000,
      timeWarning: 60000,
      // turnStrategy: '',
      gameDate: moment(new Date(), moment.ISO_8601).format(),
      showAccessCodes: true,
      logPlayerActivity: true,
      dirty: false
    },
    forces: {
      name: 'Forces',
      forces: [umpireForceTemplate],
      selectedForce: umpireForceTemplate,
      dirty: false
    },
    channels: {
      name: 'Channels',
      channels: [],
      selectedChannel: '',
      dirty: false
    },
    templates: { 
      dirty: false,
      name: 'templates',
      selectedTemplate: '',
      templates: [templateBody]
    }
  },
  wargameList: [],
  wargameInitiated: false,
  gameTurn: 0,
  phase: ADJUDICATION_PHASE,
  adjudicationStartTime: moment(new Date(), moment.ISO_8601).format()
}

export const FLEX_LAYOUT_MODEL_DEFAULT: any = {
  global: {
    tabSetTabStripHeight: 45,
    tabEnableClose: false,
    tabEnableRenderOnDemand: false,
    tabSetEnableMaximize: false
  },
  borders: [],
  layout: {
    type: 'row' as string,
    weight: 100,
    children: [
    ]
  }
}
