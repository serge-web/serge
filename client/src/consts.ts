import uniqId from 'uniqid'
import moment from 'moment'
import ExpiredStorage from 'expired-storage'
import * as Globals from '@serge/config'
import { Wargame } from '@serge/custom-types'

/** name of the document used to store the initial wargame definition */
export const DEFAULT_SERVER = 'Nelson'
export const DEFAULT_PORT = '8080'

export const MSG_STORE = 'messages'
export const SERGE_INFO = 'serge_info'
export const CHAT_CHANNEL_ID = 'game-admin'

export const PLANNING_PHASE = 'planning'
export const ADJUDICATION_PHASE = 'adjudication'

export const ADMIN_ROUTE = '/serge/admin'
export const MESSAGE_TEMPLATE_ROUTE = '/serge/messageTemplates'
export const MESSAGE_CREATOR_BASE_ROUTE = '/serge/messageCreator'
export const CREATE_TEMPLATE_ROUTE = '/create/template'
export const EDIT_TEMPLATE_ROUTE = '/edit/template'
export const CREATE_MESSAGE_ROUTE = '/create/message'
export const EDIT_MESSAGE_ROUTE = '/edit/message'
export const GAME_SETUP_ROUTE = '/serge/gameSetup'
export const WELCOME_SCREEN_EDIT_ROUTE = '/serge/editWelcomeScreen'
export const EXPORT_ROUTE = '/serge/export'
export const EXPORT_MESSAGES_SUBROUTE = '/messages'
export const EXPORT_FORCES_SUBROUTE = '/foeces'
export const EXPORT_PRINT_SUBROUTE = '/print/:id'
export const PLAYERUI_ROUTE = '/serge/player'
export const DEMO_ROUTE = '/serge/demo'
export const STORYBOOK_ROUTE = './storybook'

export const expiredStorage = new ExpiredStorage()
export const LOCAL_STORAGE_TIMEOUT = 2592000 // one month

export const MAX_LISTENERS = 82

export const UMPIRE_FORCE = 'umpire'

export const LOCATION_PENDING = 'LocationPending' // special state where platforms can be moved at turn zero

// return states of adjudication form
export const PLAN_ACCEPTED = 'accepted'
export const PLAN_REJECTED = 'rejected'

// names of special channels
export const CHANNEL_RFI_STATUS = 'rfis'
export const CHANNEL_MAPPING = 'mapping'

// NOTE: time period to wait if server returns an error. One frequent cause of error
// during development is that the server is stopped.  We're introducing a
// throttle value to prevent the browser going into a race condition
// as it sends 1000s of requests to the server
export const ERROR_THROTTLE = 3000

// Interval for check server heartbeats
export const SERVER_PING_INTERVAL = 20000

// Nov 2019. Ian modified the server path to use the
// current URL, so we can use Heroku to provide
// review instances of the app.  In these
// review instances, we can't predict the URL, so
// were failing CORS test
export const baseUrl = () => {
  const { hostname, protocol } = window.location
  // @ts-ignore
  const host = (new URL(window.location)).searchParams.get('host')

  // NOTE: for all non-heroku deployments, we need to append the port number
  const portDetails = hostname.toLowerCase().indexOf('herokuapp') !== -1 ? '' : ':' + DEFAULT_PORT

  const res = host || `${protocol}//${hostname}` + portDetails

  return res
}

export const serverPath = (
  // @ts-ignore
  window.G_CONFIG.REACT_APP_SERVER_PATH || process.env.REACT_APP_SERVER_PATH || baseUrl() + '/'
).replace(/\/?$/, '/')

export const databasePath = `${serverPath}db/`

export const headers = {
  'Content-Type': 'application/json',
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'POST',
  'Access-Control-Allow-Headers': 'Authorization, Lang'
}

export const defaultGameInfo = {
  imageUrlSet: false,
  imageUrl: '/default_img/sergeDefault.png',
  title: 'Serge',
  description: `You have arrived at the Development Centre Gaming Facility.\n
  You will use this web-based application to interact with players from other forces, together with the umpires in the White Cell.\n
  At any point during your time here you can submit insights via the Insights button at the top-right of the gaming page.\n
  These insights could relate to the current doctrine being explored, the performance of your force, or how the game is being organised / facilitated.\n
  Thanks in advance for your participation.\n
  Maj Duncan Dare, PO1 Gaming`,
  showAccessCodes: true
}

export const forceTemplate = {
  name: '',
  uniqid: '',
  overview: 'An overview written here..',
  roles: [
    {
      name: 'CO',
      roleId: `p${uniqId.time()}`,
      isGameControl: false,
      isObserver: false,
      isInsightViewer: false
    }
  ],
  iconURL: serverPath + 'default_img/forceDefault.png',
  color: '#3dd0ff',
  umpire: false,
  dirty: false
}

export const umpireForceTemplate = {
  name: 'White',
  uniqid: 'umpire',
  overview: 'Umpire force.',
  roles: [
    {
      name: 'Game Control',
      roleId: `p${uniqId.time()}`,
      isGameControl: true,
      isObserver: true,
      isInsightViewer: true
    }
  ],
  iconURL: serverPath + 'default_img/umpireDefault.png',
  color: '#FCFBEE',
  umpire: true,
  dirty: false
}

export const channelTemplate = {
  name: '',
  uniqid: '',
  participants: []
}

export const dbDefaultPlaylogSettings = {
  wargame: 'missing',
  role: 'missing',
  activityType: {
    aType: 'pending'
  },
  activityTime: '',
  _id: new Date().toISOString(), 
  isOpen: false
}

export const dbDefaultSettings: Wargame = {
  _id: Globals.wargameSettings,
  wargameTitle: '',
  name: '',
  data: {
    overview: {
      name: 'Overview - settings',
      gameDescription: '',
      // spatialRepresentation: '',
      gameTurnTime: {
        unit: 'millis',
        millis: 43200000
      },
      realtimeTurnTime: 300000,
      timeWarning: 60000,
      // turnStrategy: '',
      gameDate: moment(new Date(), moment.ISO_8601).format(),
      showAccessCodes: true,
      logPlayerActivity: true,
      dirty: false
    },
    forces: {
      name: 'Forces',
      forces: [umpireForceTemplate],
      selectedForce: umpireForceTemplate,
      dirty: false
    },
    channels: {
      name: 'Channels',
      channels: [],
      selectedChannel: '',
      dirty: false
    }
  },
  wargameList: [],
  wargameInitiated: false,
  gameTurn: 0,
  phase: ADJUDICATION_PHASE,
  adjudicationStartTime: moment(new Date(), moment.ISO_8601).format()
}

export const FLEX_LAYOUT_MODEL_DEFAULT: any = {
  global: {
    tabSetTabStripHeight: 45,
    tabEnableClose: false,
    tabEnableRenderOnDemand: false
  },
  borders: [],
  layout: {
    type: 'row' as string,
    weight: 100,
    children: [
    ]
  }
}
