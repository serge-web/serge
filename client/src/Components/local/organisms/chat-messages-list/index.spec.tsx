/* global it expect */

import { CHAT_MESSAGE } from 'src/config'
import React from 'react'
import renderer from 'react-test-renderer'
import { ChatMessage } from 'src/custom-types'

import ChatMessagesList from './index'
import moment from 'moment-timezone'

describe('ChatMessagesList renders correctly', () => {
  it('renders component correctly', () => {
    moment.tz.setDefault('Etc/UTC')
    const icons = [
      './images/default_img/forceDefault.png'
    ]
    const colors = [
      '#F00'
    ]
    const names = [
      'Blue'
    ]
    const markAllAsRead = (): void => window.alert('Callback on mark all as read')
    const messages: ChatMessage[] = [{
      messageType: CHAT_MESSAGE,
      message: { content: 'Content of message' },
      details: {
        channel: 'game-admin',
        from: {
          force: 'White',
          forceColor: '#FCFBEE',
          roleId: 'r345',
          roleName: 'Game Control',
          iconURL: 'default_img/umpireDefault.png'
        },
        messageType: 'Chat',
        timestamp: '2020-09-18T05:41:17.349Z',
        privateMessage: 'Private weather message',
        turnNumber: 1
      },
      _id: '2020-09-18T05:41:17.349Z'
    }]
    const tree = renderer
      .create(<ChatMessagesList hideForcesInChannel={false} isUmpire={true} names={names} playerRole={'r345'} playerForce={'Blue'} messages={messages} colors={colors} icons={icons} onMarkAllAsRead={markAllAsRead} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
