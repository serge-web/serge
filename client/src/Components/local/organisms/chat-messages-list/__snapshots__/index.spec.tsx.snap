// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessagesList renders correctly renders component correctly 1`] = `
<div>
  <div
    className="MuiBox-root MuiBox-root-1"
  >
    <div
      className="forces-in-channel"
    >
      <img
        alt=""
        className="force-indicator role-icon"
        src="http://localhost:8080/default_img/forceDefault.png"
        style={
          Object {
            "backgroundColor": "#F00",
            "borderColor": "#F00",
          }
        }
        title="Blue"
      />
      <button
        aria-describedby={null}
        className="MuiButtonBase-root MuiButton-root MuiButton-text mark-all-read"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Mark All as Read"
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <svg
            className="svg-inline--fa fa-envelope"
          />
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    className="MuiBox-root MuiBox-root-2 messages-list"
    style={
      Object {
        "height": "calc(100vh - 280px)",
      }
    }
  >
    <div
      className="MuiBox-root MuiBox-root-3"
    >
      <div
        className="MuiBox-root MuiBox-root-4"
        style={
          Object {
            "float": "right",
          }
        }
      >
        <div
          className="chat-message-wrapper 
      chat-message-wrapper__other"
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <span
            className="message-bar"
            onClick={[Function]}
            style={
              Object {
                "background": "#FCFBEE",
                "borderRadius": "8px 0 0 8px",
                "left": "0%",
              }
            }
          />
          <div
            className="message-text"
          >
            Content of message
          </div>
          <div
            className="MuiBox-root MuiBox-root-5"
          >
            <span>
              <span
                className="icon-private"
              >
                <svg
                  className="svg-inline--fa fa-user-secret"
                />
              </span>
              <span
                className="label-private"
              >
                Private message:
              </span>
            </span>
            <div
              className="MuiBox-root MuiBox-root-6"
            >
              <div
                className="MuiChip-root makeStyles-root-7 makeStyles-root-10 MuiChip-sizeSmall makeStyles-sizeSmall-8"
                onKeyDown={[Function]}
                onKeyUp={[Function]}
              >
                <span
                  className="MuiChip-label makeStyles-label-9 makeStyles-label-11 MuiChip-labelSmall"
                >
                  Game Control
                </span>
              </div>
              <span
                className="info-body info-body__other"
              >
                05:41
              </span>
            </div>
          </div>
          <div
            className="MuiBox-root MuiBox-root-12 private"
          >
            <div
              className="paragraph"
            >
              Private weather message
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
